src/control/server/harness.go

```diff
@@ -260,39 +252,57 @@ func (h *IOServerHarness) Start(parent context.Context, membership *system.Membe
 			}
 		}
 
-		for range instances {
+		h.log.Debug("monitoring instances")
+	Monitor:
+		for {
 			select {
 			case <-parent.Done():
+				h.log.Debug("stopping harness")
 				return parent.Err()
 			case err := <-errChan:
 				// TODO: Restart failed instances rather than shutting everything
 				// down.
-				h.log.Infof("instance exited: %v", err)
-				for _, instance := range instances {
-					h.log.Debugf("instance %d running? %v\n", instance.Index(),
-						instance.IsStarted())
+				allInstancesStopped := !h.HasStartedInstances()
+				msg := fmt.Sprintf("instance exited: %v", err)
+				if allInstancesStopped {
+					msg += ", all instances stopped!"
+					atomic.StoreInt32(&h.readyForRestart, 1)
 				}
+				h.log.Info(msg)
+			case <-h.startInstances:
+				h.log.Debug("restart instances signal received")
+				if h.HasStartedInstances() {
+					return errors.New("attempts to restart when instances are already running")
+				}
+				atomic.StoreInt32(&h.readyForRestart, 0)
```

Comment:
Realized after spending some time poring over this that it doesn't seem anyone is actually reading this readyForRestart boolean. If I'm being honest, it's a little hard to follow the expected interaction among the threads. It would be really nice to have some kind of diagram or flowchart of how this all works together, and the actual flow of each method that gets us to each of these cases.

Reply:
the readyForRestart is going to be used in the start logic in https://github.com/daos-stack/daos/pull/1541 . will add some diagram/detail to the ticket.